src/coreclr/jit/lower.cpp

```diff
@@ -7004,6 +7022,181 @@ void Lowering::ContainCheckJTrue(GenTreeOp* node)
+        }
+#ifdef TARGET_ARM64
+        else if (child->OperIsCompare())
+        {
+            child->AsOp()->SetContained();
+
+            // Ensure the children of the compare are contained correctly.
+            child->AsOp()->gtGetOp1()->ClearContained();
+            child->AsOp()->gtGetOp2()->ClearContained();
+            ContainCheckConditionalCompare(child->AsOp());
+            *startOfChain = child;
+            return true;
+        }
+#endif
+    }
+
+    return false;
+}
+
+//------------------------------------------------------------------------
+// ContainCheckAndCompareChain : Determine if an AND is a containable chain
+//
+// Arguments:
+//    node - pointer to the node
+//
+void Lowering::ContainCheckAndChain(GenTree* tree)
+{
+    assert(tree->OperIs(GT_AND));
+
+    if (!comp->opts.OptimizationEnabled())
+    {
+        return;
+    }
+
+    // First check there is a valid chain.
+    if (IsValidCompareChain(tree->AsOp()->gtGetOp2(), tree) &&
+        IsValidCompareChain(tree->AsOp()->gtGetOp1(), tree))
+    {
+        GenTree* startOfChain = nullptr;
+
+        // To ensure ordering at code generation, Op1 and the parent can
+        // only be contained if Op2 is contained.
+        if (ContainCheckCompareChain(tree->AsOp()->gtGetOp2(), tree, &startOfChain))
+        {
+            if (ContainCheckCompareChain(tree->AsOp()->gtGetOp1(), tree, &startOfChain))
+            {
+                // If op1 is the start of a chain, then it'll be generated as a standard compare.
+                if (startOfChain != nullptr)
+                {
+                    // The earliest node in the chain will be generated as a standard compare.
+                    assert(startOfChain->OperIsCompare());
+                    startOfChain->AsOp()->gtGetOp1()->ClearContained();
+                    startOfChain->AsOp()->gtGetOp2()->ClearContained();
+                    ContainCheckCompare(startOfChain->AsOp());
+                }
+            }
+        }
+
+        JITDUMP("Lowered And chain:\n");
+        DISPTREE(tree);
+    }
+}
+
+//------------------------------------------------------------------------
+// ContainCheckSelect : determine whether the source of a select should be contained.
+//
+// Arguments:
+//    node - pointer to the node
+//
+void Lowering::ContainCheckSelect(GenTreeConditional* node)
+{
+    if (!comp->opts.OptimizationEnabled())
+    {
+        return;
+    }
+
+    // Check if the compare does not need to be generated into a register.
+    GenTree* startOfChain = nullptr;
+    ContainCheckCompareChain(node->gtCond, node, &startOfChain);
+
+    if (startOfChain != nullptr)
+    {
+        // The earliest node in the chain will be generated as a standard compare.
+        assert(startOfChain->OperIsCompare());
+        startOfChain->AsOp()->gtGetOp1()->ClearContained();
+        startOfChain->AsOp()->gtGetOp2()->ClearContained();
+        ContainCheckCompare(startOfChain->AsOp());
+    }
+}
```

Comment:
I would put this in lowerarmarch.cpp as well and condition it on arm64. The xarch implementation is going to be very different.